src/compiler/args.rs

```diff
@@ -0,0 +1,820 @@
+use std;
+use std::cmp::Ordering;
+use std::ffi::OsString;
+use std::path::PathBuf;
+
+pub type Delimiter = Option<u8>;
+
+/// Representation of a parsed argument
+#[derive(PartialEq, Clone, Debug)]
+pub enum Argument {
+    /// Unknown non-flag argument ; e.g. "foo"
+    Raw(OsString),
+    /// Unknown flag argument ; e.g. "-foo"
+    UnknownFlag(OsString),
+    /// Known flag argument ; e.g. "-bar"
+    Flag(&'static str),
+    /// Known argument with a value ; e.g. "-qux bar", where the way the
+    /// value is passed is described by the ArgDisposition<Delimiter> type.
+    WithValue(&'static str, ArgumentValue, ArgDisposition<Delimiter>),
+}
+
+/// How a value is passed to an argument with a value.
+#[derive(PartialEq, Clone, Debug)]
+pub enum ArgDisposition<T> {
+    /// As "-arg value"
+    Separated,
+    /// As "-arg value", but "-arg<delimiter>value" would be valid too
+    CanBeConcatenated(T),
+    /// As "-arg<delimiter>value", but "-arg value" would be valid too
+    CanBeSeparated(T),
+    /// As "-arg<delimiter>value"
+    Concatenated(T),
+}
+
+pub use self::ArgDisposition::*;
+
+impl<T> ArgDisposition<T> {
+    fn map<U, F>(self, f: F) -> ArgDisposition<U>
+    where
+        F: FnOnce(T) -> U,
+    {
+        match self {
+            Separated => Separated,
+            CanBeConcatenated(d) => CanBeConcatenated(f(d)),
+            CanBeSeparated(d) => CanBeSeparated(f(d)),
+            Concatenated(d) => Concatenated(f(d)),
+        }
+    }
+}
+
+impl Argument {
+    /// For arguments that allow both a concatenated or separated disposition,
+    /// normalize a parsed argument to a prefered disposition.
+    pub fn normalize(self, disposition: ArgDisposition<()>) -> Self {
+        match self {
+            Argument::WithValue(s, v, CanBeConcatenated(d)) |
+            Argument::WithValue(s, v, CanBeSeparated(d)) => {
+                Argument::WithValue(s, v, disposition.map(|_| d))
+            }
+            a => a,
+        }
+    }
+
+    pub fn to_os_string(&self) -> OsString {
+        match *self {
+            Argument::Raw(ref s) |
+            Argument::UnknownFlag(ref s) => s.clone(),
+            Argument::Flag(ref s) |
+            Argument::WithValue(ref s, _, _) => s.into(),
+        }
+    }
+
+    pub fn to_str(&self) -> Option<&'static str> {
```

Comment:
FWIW, as a data point `OsStr::to_str` doesn't allocate on unix (and returns an Option<&str>)